lib/utils/datasetState.hpp

```diff
@@ -688,4 +688,121 @@ class flagState : public datasetState {
     std::string _update_id;
 };
 
+/**
+ * @brief A dataset state that keeps the beam information of a datatset.
+ *
+ * @author James Willis
+ */
+class beamState : public datasetState {
+public:
+    /**
+     * @brief Constructor
+     * @param data  The beam information as serialized by
+     *              beamState::to_json().
+     */
+    beamState(const nlohmann::json& data) {
+        try {
+            _beams = data.get<std::vector<uint32_t>>();
+        } catch (std::exception& e) {
+            throw std::runtime_error(fmt::format(
+                fmt("beamState: Failure parsing json data ({:s}): {:s}"), data.dump(4), e.what()));
+        }
+    };
+
+    /**
+     * @brief Constructor
+     * @param beams The beam information as a vector of
+     *              beam index maps.
+
+     */
+    beamState(std::vector<uint32_t> beams) : _beams(beams){};
+
+    /**
+     * @brief Constructor
+     * @param num_beams The number of beams. The indices will end up
+     *                  running from 0 to num_beams - 1
+     */
+    beamState(size_t num_beams) : _beams(num_beams) {
+        std::iota(_beams.begin(), _beams.end(), 0);
+    }
+
+    /**
+     * @brief Get beam information (read only).
+     *
+     * @return The beam information as a vector of beam index maps.
+     */
+    const std::vector<uint32_t>& get_beams() const {
+        return _beams;
+    }
+
+private:
+    /// Serialize the data of this state in a json object
+    nlohmann::json data_to_json() const override {
+        nlohmann::json j(_beams);
+        return j;
+    }
+
+    /// Time index map of the dataset state.
+    std::vector<uint32_t> _beams;
+};
+
+/**
+ * @brief A dataset state that keeps the sub-frequency information of a datatset.
+ *
+ * @author James Willis
+ */
+class subfreqState : public datasetState {
+public:
+    /**
+     * @brief Constructor
+     * @param data  The sub-frequency information as serialized by
+     *              subfreqState::to_json().
+     */
+    subfreqState(const nlohmann::json& data) {
```

Comment:
I don't think any of the other states are much better, but it would be great to document a little better what the actual state stored in here is. I gather it's subfreq index maps, but I'm not really sure what that means.